### DISCUSSION TITLE
JSON data and Quality/Weather Calculations

### DISCUSSION THREAD

**Discussion by StarGeezerPhil** [aaronwmorris/indi-allsky#706] [General]
I'm seeing some weird results within the latest_metadata.json as I try to build a javascript front-end script to output weather detection and quality details.

Having some success:
- Utilising the target or average ADU as a correction factor for the average sqm_data value seem to give fairly consistent results.
- Using the current SQM corrected against the current ADU also seems to give fairly consistent results
- Problem: target_ADU can be very unreliable due to sudden changes
- Problem: average_SQM after day/night cycle is completely unreliable (and takes hours to get back to a realistic value)
- Problem: adu_average value is, more often than not, not set.

What time periods do the average data sets show?

Would it be possible to do the following @aaronwmorris ...
- Reset average SQM at the day/night switch
- Include an ADU dataset in a similar format to the sqm_data, e.g. "ADU_data": { "max": (max for period), "min": (min for period), "avg":  (avg for period), "setting": (target in config), "target": (current ADU target as calc.) }
- Include the time of the last day/night switch - this would be useful, i.e. if within ~30 minutes we know the exposures and ADU are settling.
- Include the manual ADU target (alongside the calculated target) for the current mode
- Calculated next mode switch (e.g. "next_cycle": (time of sunset/sunrise),) - allowing us to modify calcs around this time +/- 1 hour for example and output "Sun setting" for example.

It's only in early testing, but I've had some pretty reliable weather detection once 4+ hrs into the new cycle - like I say, the day/night cutover destroys the avg data values.

![image](https://user-images.githubusercontent.com/94135145/233783409-7f481569-1c84-49f8-abe7-a77ca436985d.png)

**Comment by StarGeezerPhil** [comment]
Ohhh, also, can I view all the SQM, ADU values at a per-image level anywhere? Is it in the database and how would I connect to it?

I use beekeeper for my postgres and mysql tables, so presumably I could connect into the Pi too?

The values would be a useful reference for me to try to tweak calculations - but, importantly would also be **incredible useful for those looking to optimise their star trail settings**.

(Sorry brain thinking now - outputting the key values in the timelapse page would also help us to review the star trails generated and update config to continually improve them. I guess max/min/avg ADU and is there a comparison value stored we could use to review the max pixel/mask threshold values - these are the two I have spend the most time trying to investigate)

Once I have something more substantial, I'll try to share the code here in case it helps others.

Getting quite happy with the live output, but quite fond of the film strip effect for the prior captures:
![image](https://user-images.githubusercontent.com/94135145/233783936-23b3cd51-c26b-4fa5-b58f-5800641adf76.png)

**Reply by aaronwmorris** [reply]
The default database is sqlite, but mysql is also supported.  You can convert your database to mysql.

https://github.com/aaronwmorris/indi-allsky/wiki/Database-conversion

**Reply by StarGeezerPhil** [reply]
Happy to just hook into what's already there if I can. Do you know the access details/port etc, same username and password as the indi-allsky interface?

Can I connect using the Pi IP and a certain port - or where is the db, can I simply open it as an sql file if it's SQlite?

**Reply by aaronwmorris** [reply]
SQLite is a non-network enabled database.  You have to access the database via file IO.

It looks like beekeeper supports SQLite, you will just have to download the database to your local system.

`/var/lib/indi-allsky/indi-allsky.sqlite`

**Reply by StarGeezerPhil** [reply]
Awesome, thanks Aaron!

Very useful indeed!

Could you tell me whether some of the above items could be added to the latest_meta.json?

**Comment by aaronwmorris** [comment]
There is a lot to digest here, but I think some of the issues here are some of the variables like adu_average and target_adu are internal values used in the exposure algorithm, but probably should not be reported externally.  The only purpose in these values is use by the algorithm to decide when the exposure needs to be changed.

For instance, adu_average is completely cleared when a new exposure deviates significantly enough from the average.  The average is not recalculated until a stable exposure is found.

I will have to take a look at the other things here as I have time.

**Reply by StarGeezerPhil** [reply]
Thanks Aaron and sorry - been tied up elsewhere.

For weather conditions, I'm finding that using the avg_SQM / ADU provides fairly reliable outputs, but frame-to-frame variations in ADU really skew this. An avg_ADU to use as a benchmark would be useful to smooth out these outliers as well as improve my estimation of "improving" vs. "deteriorating" conditions.

I'm also looking at temperature combined with high daytime ADU as another condition to help vary the expected SQM values - as the sun also skews this.
